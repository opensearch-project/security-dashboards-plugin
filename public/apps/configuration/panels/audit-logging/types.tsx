/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export interface GeneralSettings {
  // layer
  enable_rest?: boolean;
  enable_transport?: boolean;
  disabled_rest_categories?: string[];
  disabled_transport_categories?: string[];

  // attribute
  resolve_bulk_requests?: boolean;
  resolve_indices?: boolean;
  log_request_body?: boolean;
  exclude_sensitive_headers?: boolean;

  // ignore
  ignore_users?: string[];
  ignore_requests?: string[];
}

export interface ComplianceSettings {
  enabled?: boolean;

  // read
  read_metadata_only?: boolean;
  read_ignore_users?: string[];
  read_watched_fields?: {
    [key: string]: string;
  };

  // write
  write_metadata_only?: boolean;
  write_log_diffs?: boolean;
  write_ignore_users?: string[];
  write_watched_indices?: string[];
}

export interface AuditLoggingSettings {
  enabled?: boolean;
  audit?: GeneralSettings;
  compliance?: ComplianceSettings;
}
