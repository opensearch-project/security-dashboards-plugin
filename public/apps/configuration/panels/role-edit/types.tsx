/*
 *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { EuiComboBoxOptionOption } from '@elastic/eui';

export type ComboBoxOptions = EuiComboBoxOptionOption[];
export type FieldLevelSecurityMethod = 'exclude' | 'include';
export interface RoleIndexPermissionStateClass {
  indexPatterns: ComboBoxOptions;
  docLevelSecurity: string;
  fieldLevelSecurityMethod: FieldLevelSecurityMethod;
  fieldLevelSecurityFields: ComboBoxOptions;
  maskedFields: ComboBoxOptions;
  allowedActions: ComboBoxOptions;
}

export enum TenantPermissionType {
  None = '',
  Read = 'r',
  Write = 'w',
  Full = 'rw',
}

export interface RoleTenantPermissionStateClass {
  tenantPatterns: ComboBoxOptions;
  permissionType: TenantPermissionType;
}
