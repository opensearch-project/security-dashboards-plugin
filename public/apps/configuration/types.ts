/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { EuiComboBoxOptionOption } from '@elastic/eui';

export type ComboBoxOptions = EuiComboBoxOptionOption[];

export type FieldLevelSecurityMethod = 'exclude' | 'include';

export enum ResourceType {
  roles = 'roles',
  users = 'users',
  permissions = 'permissions',
  tenants = 'tenants',
  tenantsManageTab = 'tenantsManageTab',
  tenantsConfigureTab = 'tenantsConfigureTab',
  auth = 'auth',
  auditLogging = 'auditLogging',
}

export enum Action {
  view = 'view',
  create = 'create',
  edit = 'edit',
  duplicate = 'duplicate',
  manage = 'manage',
  configure = 'configure',
}

export enum SubAction {
  mapuser = 'mapuser',
}

export enum PageId {
  dashboardId = 'dashboards',
  visualizationId = 'visualize',
}

export enum TenantPermissionType {
  None = '',
  Read = 'Read only',
  ReadWrite = 'Read and Write',
}

export interface RouteItem {
  name: string;
  href: string;
}

export interface DataObject<T> {
  [key: string]: T;
}

export interface ObjectsMessage<T> {
  total: number;
  data: DataObject<T>;
}

export interface RoleIndexPermission {
  index_patterns: string[];
  dls: string;
  fls: string[];
  masked_fields: string[];
  allowed_actions: string[];
}

export interface RoleIndexPermissionView extends RoleIndexPermission {
  id: number;
}

export interface RoleTenantPatterns {
  tenant_patterns: string[];
}

export interface RoleTenantPermission extends RoleTenantPatterns {
  allowed_actions: string[];
}

export interface RoleTenantPermissionView extends RoleTenantPatterns {
  permissionType: string;
}

export interface RoleUpdate {
  cluster_permissions: string[];
  index_permissions: RoleIndexPermission[];
  tenant_permissions: RoleTenantPermission[];
}

export interface RoleDetail extends RoleUpdate {
  reserved: boolean;
}

export interface RoleMappingDetail {
  backend_roles: string[];
  hosts: string[];
  users: string[];
}

export interface TenantUpdate {
  description: string;
}

export interface TenantName {
  tenant: string;
}

export interface Tenant extends TenantUpdate, TenantName {
  reserved: boolean;
  tenantValue: string;
}

export interface TenantSelect extends TenantName {
  username: string;
}

export interface RoleTenantPermissionDetail extends RoleTenantPermissionView, Tenant {}

export interface UserAttributes {
  [key: string]: string;
}

export interface InternalUser {
  attributes: UserAttributes;
  backend_roles: string[];
}

export interface InternalUserUpdate extends InternalUser {
  password?: string;
}

export interface ActionGroupUpdate {
  allowed_actions: string[];
}

export interface ActionGroupItem extends ActionGroupUpdate {
  reserved: boolean;
  type: 'cluster' | 'index' | 'all' | 'kibana' | undefined;
}

export interface ExpandedRowMapInterface {
  [key: string]: React.ReactNode;
}

export interface FormRowDeps {
  headerText: string;
  optional?: boolean;
  headerSubText?: string;
  helpLink?: string;
  helpText?: string;
}
