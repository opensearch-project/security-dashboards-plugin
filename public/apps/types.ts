import { AppMountParameters, CoreStart } from '../../../../src/core/public';
import { AppPluginStartDependencies } from '../types';

export interface AppDependencies {
    coreStart: CoreStart,
    navigation: AppPluginStartDependencies,
    params: AppMountParameters
}
