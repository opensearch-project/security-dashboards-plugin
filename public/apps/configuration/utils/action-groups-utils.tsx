/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { HttpStart } from 'opensearch-dashboards/public';
import { map } from 'lodash';
import { API_ENDPOINT_ACTIONGROUPS, CLUSTER_PERMISSIONS, INDEX_PERMISSIONS } from '../constants';
import { DataObject, ActionGroupItem, ActionGroupUpdate, ObjectsMessage } from '../types';
import { httpDelete, httpGet, httpPost } from './request-utils';
import { getResourceUrl } from './resource-utils';

export interface PermissionListingItem {
  name: string;
  type: 'Action group' | 'Single permission';
  reserved: boolean;
  allowedActions: string[];
  hasClusterPermission: boolean;
  hasIndexPermission: boolean;
}

export async function fetchActionGroups(http: HttpStart): Promise<DataObject<ActionGroupItem>> {
  const actiongroups = await httpGet<ObjectsMessage<ActionGroupItem>>({
    http,
    url: API_ENDPOINT_ACTIONGROUPS,
  });
  return actiongroups.data;
}

export function transformActionGroupsToListingFormat(
  rawData: DataObject<ActionGroupItem>
): PermissionListingItem[] {
  return map(rawData, (value: ActionGroupItem, key?: string) => ({
    name: key || '',
    type: 'Action group',
    reserved: value.reserved,
    allowedActions: value.allowed_actions,
    hasClusterPermission: value.type === 'cluster' || value.type === 'all',
    hasIndexPermission: value.type === 'index' || value.type === 'all',
  }));
}

export async function fetchActionGroupListing(http: HttpStart): Promise<PermissionListingItem[]> {
  return transformActionGroupsToListingFormat(await fetchActionGroups(http));
}

function getClusterSinglePermissions(): PermissionListingItem[] {
  return CLUSTER_PERMISSIONS.map((permission) => ({
    name: permission,
    type: 'Single permission',
    reserved: true,
    allowedActions: [],
    hasClusterPermission: true,
    hasIndexPermission: false,
  }));
}

function getIndexSinglePermissions(): PermissionListingItem[] {
  return INDEX_PERMISSIONS.map((permission) => ({
    name: permission,
    type: 'Single permission',
    reserved: true,
    allowedActions: [],
    hasClusterPermission: false,
    hasIndexPermission: true,
  }));
}

export async function getAllPermissionsListing(http: HttpStart): Promise<PermissionListingItem[]> {
  return mergeAllPermissions(await fetchActionGroups(http));
}

export async function mergeAllPermissions(
  actionGroups: DataObject<ActionGroupItem>
): Promise<PermissionListingItem[]> {
  return transformActionGroupsToListingFormat(actionGroups)
    .concat(getClusterSinglePermissions())
    .concat(getIndexSinglePermissions());
}

export async function updateActionGroup(
  http: HttpStart,
  groupName: string,
  updateObject: ActionGroupUpdate
): Promise<ActionGroupUpdate> {
  return await httpPost({
    http,
    url: getResourceUrl(API_ENDPOINT_ACTIONGROUPS, groupName),
    body: updateObject,
  });
}

export async function requestDeleteActionGroups(http: HttpStart, groups: string[]) {
  for (const group of groups) {
    await httpDelete({ http, url: getResourceUrl(API_ENDPOINT_ACTIONGROUPS, group) });
  }
}
