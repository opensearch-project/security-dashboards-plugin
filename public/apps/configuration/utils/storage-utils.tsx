/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { Toast } from '@elastic/eui/src/components/toast/global_toast_list';

/*
 * This file leverage storage for frontend data. Currently it is for cross page toasts only,
 * so session storage is in use (no need to hold the toast message to until next session)
 */

export function getValue<T>(category: string, key: string): T | null {
  const item = sessionStorage.getItem(`${category}::${key}`);
  if (!item) {
    return null;
  }
  return JSON.parse(item) as T;
}

export function setValue(category: string, key: string, value: any) {
  sessionStorage.setItem(`${category}::${key}`, JSON.stringify(value));
}

const TOAST_CATEGORY = 'toast';

export function setCrossPageToast(targetUrl: string, toast: Toast) {
  setValue(TOAST_CATEGORY, targetUrl, toast);
}

export function getAndClearCrossPageToast(targetUrl: string): Toast | null {
  const result = getValue<Toast>(TOAST_CATEGORY, targetUrl);
  setValue(TOAST_CATEGORY, targetUrl, null);
  return result;
}
