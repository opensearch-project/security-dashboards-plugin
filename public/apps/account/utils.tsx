/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { HttpStart } from 'opensearch-dashboards/public';
import { API_AUTH_LOGOUT, OPENID_AUTH_LOGOUT, SAML_AUTH_LOGOUT } from '../../../common';
import { API_ENDPOINT_ACCOUNT_INFO } from './constants';
import { AccountInfo } from './types';
import { httpGet, httpGetWithIgnores, httpPost } from '../configuration/utils/request-utils';
import { setShouldShowTenantPopup } from '../../utils/storage-utils';

export function fetchAccountInfo(http: HttpStart): Promise<AccountInfo> {
  return httpGet(http, API_ENDPOINT_ACCOUNT_INFO);
}

export async function fetchAccountInfoSafe(http: HttpStart): Promise<AccountInfo | undefined> {
  return httpGetWithIgnores<AccountInfo>(http, API_ENDPOINT_ACCOUNT_INFO, [401]);
}

export async function logout(http: HttpStart, logoutUrl?: string): Promise<void> {
  await httpPost(http, API_AUTH_LOGOUT);
  setShouldShowTenantPopup(null);
  // Clear everything in the sessionStorage since they can contain sensitive information
  sessionStorage.clear();
  // When no basepath is set, we can take '/' as the basepath.
  const basePath = http.basePath.serverBasePath ? http.basePath.serverBasePath : '/';
  const nextUrl = encodeURIComponent(basePath);
  window.location.href =
    logoutUrl || `${http.basePath.serverBasePath}/app/login?nextUrl=${nextUrl}`;
}

export async function samlLogout(http: HttpStart): Promise<void> {
  // This will ensure tenancy is picked up from local storage in the next login.
  setShouldShowTenantPopup(null);
  window.location.href = `${http.basePath.serverBasePath}${SAML_AUTH_LOGOUT}`;
}

export async function openidLogout(http: HttpStart): Promise<void> {
  // This will ensure tenancy is picked up from local storage in the next login.
  setShouldShowTenantPopup(null);
  sessionStorage.clear();
  window.location.href = `${http.basePath.serverBasePath}${OPENID_AUTH_LOGOUT}`;
}

export async function externalLogout(http: HttpStart, logoutEndpoint: string): Promise<void> {
  // This will ensure tenancy is picked up from local storage in the next login.
  setShouldShowTenantPopup(null);
  sessionStorage.clear();
  window.location.href = `${http.basePath.serverBasePath}${logoutEndpoint}`;
}

export async function updateNewPassword(
  http: HttpStart,
  newPassword: string,
  currentPassword: string
): Promise<void> {
  await httpPost(http, API_ENDPOINT_ACCOUNT_INFO, {
    password: newPassword,
    current_password: currentPassword,
  });
}
