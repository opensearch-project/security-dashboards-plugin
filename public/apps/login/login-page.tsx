/*
 *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React, { useState } from 'react';
import {
  EuiText,
  EuiFieldText,
  EuiIcon,
  EuiSpacer,
  EuiButton,
  EuiImage,
  EuiListGroup,
  EuiForm,
  EuiFormRow,
} from '@elastic/eui';
import { CoreStart } from '../../../../../src/core/public';
import { ClientConfigType } from '../../types';
import defaultBrandImage from '../../assets/open_distro_for_elasticsearch_logo_h.svg';
import { validateCurrentPassword } from '../../utils/login-utils';

interface LoginPageDeps {
  http: CoreStart['http'];
  config: ClientConfigType['ui']['basicauth']['login'];
}

function redirect(serverBasePath: string) {
  // navigate to nextUrl
  const urlParams = new URLSearchParams(window.location.search);
  let nextUrl = urlParams.get('nextUrl');
  if (!nextUrl || nextUrl.toLowerCase().includes('//')) {
    nextUrl = serverBasePath;
  }
  window.location.href = nextUrl + window.location.hash;
}

export function LoginPage(props: LoginPageDeps) {
  const [username, setUsername] = React.useState('');
  const [password, setPassword] = React.useState('');
  const [loginFailed, setloginFailed] = useState(false);
  const [usernameValidationFailed, setUsernameValidationFailed] = useState(false);
  const [passwordValidationFailed, setPasswordValidationFailed] = useState(false);

  let errorLabel = null;
  if (loginFailed) {
    errorLabel = (
      <EuiText id="error" color="danger" textAlign="center">
        <b>Invalid username or password, please try again</b>
      </EuiText>
    );
  }

  // @ts-ignore : Parameter 'e' implicitly has an 'any' type.
  const handleSubmit = async (e) => {
    e.preventDefault();

    // Clear errors
    setloginFailed(false);
    setUsernameValidationFailed(false);
    setPasswordValidationFailed(false);

    // Form validation
    if (username === '') {
      setUsernameValidationFailed(true);
      return;
    }

    if (password === '') {
      setPasswordValidationFailed(true);
      return;
    }

    try {
      await validateCurrentPassword(props.http, username, password);
      redirect(props.http.basePath.serverBasePath);
    } catch (error) {
      console.log(error);
      setloginFailed(true);
      return;
    }
  };

  // TODO: Get brand image from server config
  return (
    <EuiListGroup className="login-wrapper">
      {props.config.showbrandimage && (
        <EuiImage size="fullWidth" alt="" url={props.config.brandimage || defaultBrandImage} />
      )}
      <EuiSpacer size="s" />
      <EuiText size="m" textAlign="center">
        {props.config.title || 'Please login to Kibana'}
      </EuiText>
      <EuiSpacer size="s" />
      <EuiText size="s" textAlign="center">
        {props.config.subtitle ||
          'If you have forgotten your username or password, please ask your system administrator'}
      </EuiText>
      <EuiSpacer size="s" />
      <EuiForm component="form">
        <EuiFormRow>
          <EuiFieldText
            data-test-subj="user-name"
            placeholder="Username"
            prepend={<EuiIcon type="user" />}
            onChange={(e) => setUsername(e.target.value)}
            value={username}
            isInvalid={usernameValidationFailed}
          />
        </EuiFormRow>
        <EuiFormRow isInvalid={passwordValidationFailed}>
          <EuiFieldText
            data-test-subj="password"
            placeholder="Password"
            prepend={<EuiIcon type="lock" />}
            type="password"
            onChange={(e) => setPassword(e.target.value)}
            value={password}
            isInvalid={usernameValidationFailed}
          />
        </EuiFormRow>
        <EuiFormRow>
          <EuiButton
            data-test-subj="submit"
            fill
            size="s"
            type="submit"
            className={props.config.buttonstyle || 'btn-login'}
            onClick={handleSubmit}
          >
            Log In
          </EuiButton>
        </EuiFormRow>
        {errorLabel}
      </EuiForm>
    </EuiListGroup>
  );
}
