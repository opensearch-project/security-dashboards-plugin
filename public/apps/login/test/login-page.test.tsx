/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { shallow } from 'enzyme';
import React from 'react';
import { ClientConfigType } from '../../../types';
import { LoginPage, extractNextUrlFromWindowLocation, getNextPath } from '../login-page';
import { validateCurrentPassword } from '../../../utils/login-utils';
import { API_AUTH_LOGOUT } from '../../../../common';
import { chromeServiceMock } from '../../../../../../src/core/public/mocks';
import { AuthType } from '../../../../common';
import { setSavedTenant } from '../../../utils/storage-utils';

jest.mock('../../../utils/login-utils', () => ({
  validateCurrentPassword: jest.fn(),
}));

const configUI = {
  basicauth: {
    login: {
      title: 'Title1',
      subtitle: 'SubTitle1',
      showbrandimage: true,
      brandimage: 'http://localhost:5601/images/test.png',
      buttonstyle: 'test-btn-style',
    },
  },
  openid: {
    login: {
      buttonname: 'Button1',
      showbrandimage: true,
      brandimage: 'http://localhost:5601/images/test.png',
      buttonstyle: 'test-btn-style',
    },
  },
  saml: {
    login: {
      buttonname: 'Button2',
      showbrandimage: true,
      brandimage: 'http://localhost:5601/images/test.png',
      buttonstyle: 'test-btn-style',
    },
  },
  autologout: true,
  backend_configurable: true,
};

const configUiDefault = {
  basicauth: {
    login: {
      showbrandimage: true,
    },
  },
};

describe('test extractNextUrlFromWindowLocation', () => {
  test('extract next url from window with nextUrl', () => {
    // Trick to mock window.location
    const originalLocation = window.location;
    delete window.location;
    window.location = new URL(
      "http://localhost:5601/app/login?nextUrl=%2Fapp%2Fdashboards#/view/7adfa750-4c81-11e8-b3d7-01146121b73d?_g=(filters:!(),refreshInterval:(pause:!f,value:900000),time:(from:now-24h,to:now))&_a=(description:'Analyze%20mock%20flight%20data%20for%20OpenSearch-Air,%20Logstash%20Airways,%20OpenSearch%20Dashboards%20Airlines%20and%20BeatsWest',filters:!(),fullScreenMode:!f,options:(hidePanelTitles:!f,useMargins:!t),query:(language:kuery,query:''),timeRestore:!t,title:'%5BFlights%5D%20Global%20Flight%20Dashboard',viewMode:view)"
    ) as any;
    expect(extractNextUrlFromWindowLocation()).toEqual(
      "?nextUrl=%2Fapp%2Fdashboards#/view/7adfa750-4c81-11e8-b3d7-01146121b73d?_g=(filters:!(),refreshInterval:(pause:!f,value:900000),time:(from:now-24h,to:now))&_a=(description:'Analyze%20mock%20flight%20data%20for%20OpenSearch-Air,%20Logstash%20Airways,%20OpenSearch%20Dashboards%20Airlines%20and%20BeatsWest',filters:!(),fullScreenMode:!f,options:(hidePanelTitles:!f,useMargins:!t),query:(language:kuery,query:''),timeRestore:!t,title:'%5BFlights%5D%20Global%20Flight%20Dashboard',viewMode:view)"
    );
  });

  test('extract next url from window without nextUrl', () => {
    const originalLocation = window.location;
    delete window.location;
    window.location = new URL('http://localhost:5601/app/home');
    expect(extractNextUrlFromWindowLocation()).toEqual('?nextUrl=%2F');
  });
});

describe('test redirect', () => {
  test('extract redirect excludes security_tenant when no tenant in local storage', () => {
    // Trick to mock window.location
    const originalLocation = window.location;
    delete window.location;
    window.location = new URL('http://localhost:5601/app/login?nextUrl=%2Fapp%2Fdashboards') as any;
    setSavedTenant(null);
    const nextPath = getNextPath('');
    expect(nextPath).toEqual('/app/dashboards');
    window.location = originalLocation;
  });

  test('extract redirect includes security_tenant when tenant in local storage', () => {
    const originalLocation = window.location;
    delete window.location;
    window.location = new URL('http://localhost:5601/app/login?nextUrl=%2Fapp%2Fdashboards');
    setSavedTenant('custom');
    const nextPath = getNextPath('');
    expect(nextPath).toEqual('/app/dashboards?security_tenant=custom');
    setSavedTenant(null);
    window.location = originalLocation;
  });

  test('extract redirect includes security_tenant when tenant in local storage, existing url params and hash', () => {
    const originalLocation = window.location;
    delete window.location;
    window.location = new URL(
      "http://localhost:5601/app/login?nextUrl=%2Fapp%2Fdashboards?param1=value1#/view/7adfa750-4c81-11e8-b3d7-01146121b73d?_g=(filters:!(),refreshInterval:(pause:!f,value:900000),time:(from:now-24h,to:now))&_a=(description:'Analyze%20mock%20flight%20data%20for%20OpenSearch-Air,%20Logstash%20Airways,%20OpenSearch%20Dashboards%20Airlines%20and%20BeatsWest',filters:!(),fullScreenMode:!f,options:(hidePanelTitles:!f,useMargins:!t),query:(language:kuery,query:''),timeRestore:!t,title:'%5BFlights%5D%20Global%20Flight%20Dashboard',viewMode:view)"
    );
    setSavedTenant('custom');
    const nextPath = getNextPath('');
    expect(nextPath).toEqual(
      "/app/dashboards?param1=value1&security_tenant=custom#/view/7adfa750-4c81-11e8-b3d7-01146121b73d?_g=(filters:!(),refreshInterval:(pause:!f,value:900000),time:(from:now-24h,to:now))&_a=(description:'Analyze%20mock%20flight%20data%20for%20OpenSearch-Air,%20Logstash%20Airways,%20OpenSearch%20Dashboards%20Airlines%20and%20BeatsWest',filters:!(),fullScreenMode:!f,options:(hidePanelTitles:!f,useMargins:!t),query:(language:kuery,query:''),timeRestore:!t,title:'%5BFlights%5D%20Global%20Flight%20Dashboard',viewMode:view)"
    );
    setSavedTenant(null);
    window.location = originalLocation;
  });
});

describe('Login page', () => {
  let chrome: ReturnType<typeof chromeServiceMock.createStartContract>;
  const mockHttpStart = {
    basePath: {
      serverBasePath: '/app/opensearch-dashboards',
    },
  };

  beforeEach(() => {
    chrome = chromeServiceMock.createStartContract();
  });

  describe('renders', () => {
    it('renders with config value: string array', () => {
      const config: ClientConfigType = {
        ui: configUI,
        auth: {
          type: [AuthType.BASIC],
          logout_url: API_AUTH_LOGOUT,
        },
      };
      const component = shallow(
        <LoginPage http={mockHttpStart as any} chrome={chrome} config={config as any} />
      );
      expect(component).toMatchSnapshot();
    });

    it('renders with config value: string', () => {
      const config: ClientConfigType = {
        ui: configUI,
        auth: {
          type: AuthType.BASIC,
          logout_url: API_AUTH_LOGOUT,
        },
      };
      const component = shallow(
        <LoginPage http={mockHttpStart as any} chrome={chrome} config={config as any} />
      );
      expect(component).toMatchSnapshot();
    });

    it('renders with config value for multiauth', () => {
      const config: ClientConfigType = {
        ui: configUI,
        auth: {
          type: [AuthType.BASIC, 'openid', AuthType.SAML],
          logout_url: API_AUTH_LOGOUT,
        },
      };
      const component = shallow(
        <LoginPage http={mockHttpStart as any} chrome={chrome} config={config as any} />
      );
      expect(component).toMatchSnapshot();
    });

    it('renders with default value: string array', () => {
      const config: ClientConfigType = {
        ui: configUiDefault,
        auth: {
          type: [''],
        },
      };
      const component = shallow(
        <LoginPage http={mockHttpStart as any} chrome={chrome} config={config as any} />
      );
      expect(component).toMatchSnapshot();
    });

    it('renders with default value: string', () => {
      const config: ClientConfigType = {
        ui: configUiDefault,
        auth: {
          type: '',
        },
      };
      const component = shallow(
        <LoginPage http={mockHttpStart as any} chrome={chrome} config={config as any} />
      );
      expect(component).toMatchSnapshot();
    });
  });

  describe('event trigger testing', () => {
    let component;
    const setState = jest.fn();
    const useState = jest.spyOn(React, 'useState');
    const config: ClientConfigType = {
      ui: configUiDefault,
      auth: {
        type: AuthType.BASIC,
      },
    };
    beforeEach(() => {
      useState.mockImplementation((initialValue) => [initialValue, setState]);
      component = shallow(
        <LoginPage http={mockHttpStart as any} chrome={chrome} config={config as any} />
      );
    });

    it('should update user name field on change event', () => {
      const event = {
        target: { value: 'dummy' },
      } as React.ChangeEvent<HTMLInputElement>;
      component.find('[data-test-subj="user-name"]').simulate('change', event);
      expect(setState).toBeCalledWith('dummy');
    });

    it('should update password field on change event', () => {
      const event = {
        target: { value: 'dummy' },
      } as React.ChangeEvent<HTMLInputElement>;
      component.find('[data-test-subj="password"]').simulate('change', event);
      expect(setState).toBeCalledWith('dummy');
    });
  });

  describe('handle submit event', () => {
    let component;
    const useState = jest.spyOn(React, 'useState');
    const setState = jest.fn();
    const config: ClientConfigType = {
      ui: configUiDefault,
      auth: {
        type: AuthType.BASIC,
      },
    };
    beforeEach(() => {
      useState.mockImplementation(() => ['user1', setState]);
      useState.mockImplementation(() => ['password1', setState]);
      component = shallow(
        <LoginPage http={mockHttpStart as any} chrome={chrome} config={config as any} />
      );
    });

    it('submit click event', () => {
      window = Object.create(window);
      const url = 'http://dummy.com';
      Object.defineProperty(window, 'location', {
        value: {
          href: url,
        },
      });
      component.find('[data-test-subj="submit"]').simulate('click', {
        preventDefault: () => {},
      });
      expect(validateCurrentPassword).toHaveBeenCalledTimes(1);
    });
  });
});
