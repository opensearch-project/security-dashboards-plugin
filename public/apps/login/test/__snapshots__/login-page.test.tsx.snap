// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page renders renders with config value 1`] = `
<EuiListGroup
  className="login-wrapper"
>
  <EuiImage
    alt=""
    size="fullWidth"
    url="http://localhost:5601/images/test.png"
  />
  <EuiSpacer
    size="s"
  />
  <EuiText
    size="m"
    textAlign="center"
  >
    Title1
  </EuiText>
  <EuiSpacer
    size="s"
  />
  <EuiText
    size="s"
    textAlign="center"
  >
    SubTitle1
  </EuiText>
  <EuiSpacer
    size="s"
  />
  <EuiForm
    component="form"
  >
    <EuiFormRow
      describedByIds={Array []}
      display="row"
      fullWidth={false}
      hasChildLabel={true}
      hasEmptyLabelSpace={false}
      labelType="label"
    >
      <EuiFieldText
        data-test-subj="user-name"
        isInvalid={false}
        onChange={[Function]}
        placeholder="Username"
        prepend={
          <EuiIcon
            type="user"
          />
        }
        value=""
      />
    </EuiFormRow>
    <EuiFormRow
      describedByIds={Array []}
      display="row"
      fullWidth={false}
      hasChildLabel={true}
      hasEmptyLabelSpace={false}
      isInvalid={false}
      labelType="label"
    >
      <EuiFieldText
        data-test-subj="password"
        isInvalid={false}
        onChange={[Function]}
        placeholder="Password"
        prepend={
          <EuiIcon
            type="lock"
          />
        }
        type="password"
        value=""
      />
    </EuiFormRow>
    <EuiFormRow
      describedByIds={Array []}
      display="row"
      fullWidth={false}
      hasChildLabel={true}
      hasEmptyLabelSpace={false}
      labelType="label"
    >
      <EuiButton
        className="test-btn-style"
        data-test-subj="submit"
        fill={true}
        onClick={[Function]}
        size="s"
        type="submit"
      >
        Log In
      </EuiButton>
    </EuiFormRow>
  </EuiForm>
</EuiListGroup>
`;

exports[`Login page renders renders with default value 1`] = `
<EuiListGroup
  className="login-wrapper"
>
  <EuiSpacer
    size="s"
  />
  <EuiText
    size="m"
    textAlign="center"
  >
    Please login to Kibana
  </EuiText>
  <EuiSpacer
    size="s"
  />
  <EuiText
    size="s"
    textAlign="center"
  >
    If you have forgotten your username or password, please ask your system administrator
  </EuiText>
  <EuiSpacer
    size="s"
  />
  <EuiForm
    component="form"
  >
    <EuiFormRow
      describedByIds={Array []}
      display="row"
      fullWidth={false}
      hasChildLabel={true}
      hasEmptyLabelSpace={false}
      labelType="label"
    >
      <EuiFieldText
        data-test-subj="user-name"
        isInvalid={false}
        onChange={[Function]}
        placeholder="Username"
        prepend={
          <EuiIcon
            type="user"
          />
        }
        value=""
      />
    </EuiFormRow>
    <EuiFormRow
      describedByIds={Array []}
      display="row"
      fullWidth={false}
      hasChildLabel={true}
      hasEmptyLabelSpace={false}
      isInvalid={false}
      labelType="label"
    >
      <EuiFieldText
        data-test-subj="password"
        isInvalid={false}
        onChange={[Function]}
        placeholder="Password"
        prepend={
          <EuiIcon
            type="lock"
          />
        }
        type="password"
        value=""
      />
    </EuiFormRow>
    <EuiFormRow
      describedByIds={Array []}
      display="row"
      fullWidth={false}
      hasChildLabel={true}
      hasEmptyLabelSpace={false}
      labelType="label"
    >
      <EuiButton
        className="btn-login"
        data-test-subj="submit"
        fill={true}
        onClick={[Function]}
        size="s"
        type="submit"
      >
        Log In
      </EuiButton>
    </EuiFormRow>
  </EuiForm>
</EuiListGroup>
`;
